Automattic/woocommerce-payments
# Add VAT details form when downloading invoice

Fixes #3934 

#### Changes proposed in this Pull Request

This PR builds on top of #4140, adding the actual VAT details form to the modal. The form is composed of two steps: in the first, we ask merchants for their VAT number — in case they have one — and validate it; in the second step we use the data we were able to get when validating the VAT number to ask for the company name and address.

Merchants are able to submit the details without specifying a VAT number, but business name and address are required. Once the details are stored in the server, the `hasSubmittedVatData` flag is set to true on the client to allow invoices to be downloaded without requiring a page refresh. The account cache is also refreshed in the background, so that the new status is reflected right away, instead of having to wait for an automatic refresh.

As soon as the form is submitted and the flag is updated, merchants will be redirected to the invoice document, without having to click on Download a second time. Further downloads should be processed directly, without asking for the details again.

#### Testing instructions

_We don't have a way to validate VAT numbers for some countries, like the US. In these cases, an error will be displayed when attempting to validate it. Merchants should be able to proceed without informing a VAT number in such cases._

_If using an account from the UK, the local server and sandbox will use a test API. You can find valid VAT numbers to test it [here](https://github.com/hmrc/vat-registered-companies-api/blob/main/public/api/conf/1.0/test-data/vrn.csv)._

_If using an account from the EU, the VIES production service will be used, so you'll need to use a real VAT number._

* Apply this PR's branch and 1803-gh-Automattic/woocommerce-payments-server on your local server or sandbox
* Create a new `vat_invoice` for your account if you don't have one by running the following command
    ```
    wp wcpay add_document_for_account {Your account id} vat_invoice --date="2022-03-01" --period_from="2022-02-01 00:00:00" --period_to="2022-02-28 23:59:59"
    ```
* Navigate to the documents list and click on the download button for the VAT invoice
    * Make sure a modal pops up with the VAT details form
    * Check "I’m registered for a VAT number" and make sure your account's country prefix is used in the input
        * Input an invalid VAT number, click on Continue, and assert that a notice is displayed with the error
        * Input a valid VAT number, click on Continue, and assert that the form proceeds to the next step with the details from the VAT check
            * Modify or keep the placeholder values for business name and address and click on Confirm
            * Make sure the form is submitted correctly and the document is downloaded automatically
            * Check the account_meta table on the server to make sure the data was saved correctly
* Remove the VAT details from the server and clear the account cache on the client
    * Navigate to the documents list and click on the download button for the VAT invoice
    * Click on Continue without checking "I’m registered for a VAT number"
        * Make sure the Confirm button is disabled until a business name and address are provided
    * Provide a business name and address and click on Confirm
        * Make sure the form is submitted correctly and the document is downloaded automatically 
        * Check the account_meta table on the server to make sure the data was saved correctly

-------------------


**Post merge**
